ReactiveCocoa/Swift/Signal.swift

```diff
@@ -1309,98 +1252,325 @@ private struct ThrottleState<Value> {
 /// `combineLatestWith`.
 @warn_unused_result(message="Did you forget to call `observe` on the signal?")
 public func combineLatest<A, B, C, D, E, F, Error>(a: Signal<A, Error>, _ b: Signal<B, Error>, _ c: Signal<C, Error>, _ d: Signal<D, Error>, _ e: Signal<E, Error>, _ f: Signal<F, Error>) -> Signal<(A, B, C, D, E, F), Error> {
-	return combineLatest(a, b, c, d, e)
-		.combineLatestWith(f)
-		.map(repack)
+	
+	let collector = TupleCollector6<A,B,C,D,E,F>()
+	let states = Atomic(CombineLatestStates(collector))
+	
+	return Signal{ observer in
+		
+		let disposable = CompositeDisposable()
+		
+		disposable += a.observeForCombineLatest(states, index: 0, observer: observer)
+		disposable += b.observeForCombineLatest(states, index: 1, observer: observer)
+		disposable += c.observeForCombineLatest(states, index: 2, observer: observer)
+		disposable += d.observeForCombineLatest(states, index: 3, observer: observer)
+		disposable += e.observeForCombineLatest(states, index: 4, observer: observer)
+		disposable += f.observeForCombineLatest(states, index: 5, observer: observer)
+		
+		return disposable
+	}
 }
 
 /// Combines the values of all the given signals, in the manner described by
 /// `combineLatestWith`.
 @warn_unused_result(message="Did you forget to call `observe` on the signal?")
 public func combineLatest<A, B, C, D, E, F, G, Error>(a: Signal<A, Error>, _ b: Signal<B, Error>, _ c: Signal<C, Error>, _ d: Signal<D, Error>, _ e: Signal<E, Error>, _ f: Signal<F, Error>, _ g: Signal<G, Error>) -> Signal<(A, B, C, D, E, F, G), Error> {
-	return combineLatest(a, b, c, d, e, f)
-		.combineLatestWith(g)
-		.map(repack)
+	
+	let collector = TupleCollector7<A,B,C,D,E,F,G>()
+	let states = Atomic(CombineLatestStates(collector))
+	
+	return Signal{ observer in
+		
+		let disposable = CompositeDisposable()
+		
+		disposable += a.observeForCombineLatest(states, index: 0, observer: observer)
+		disposable += b.observeForCombineLatest(states, index: 1, observer: observer)
+		disposable += c.observeForCombineLatest(states, index: 2, observer: observer)
+		disposable += d.observeForCombineLatest(states, index: 3, observer: observer)
+		disposable += e.observeForCombineLatest(states, index: 4, observer: observer)
+		disposable += f.observeForCombineLatest(states, index: 5, observer: observer)
+		disposable += g.observeForCombineLatest(states, index: 6, observer: observer)
+
+		return disposable
+	}
 }
 
 /// Combines the values of all the given signals, in the manner described by
 /// `combineLatestWith`.
 @warn_unused_result(message="Did you forget to call `observe` on the signal?")
 public func combineLatest<A, B, C, D, E, F, G, H, Error>(a: Signal<A, Error>, _ b: Signal<B, Error>, _ c: Signal<C, Error>, _ d: Signal<D, Error>, _ e: Signal<E, Error>, _ f: Signal<F, Error>, _ g: Signal<G, Error>, _ h: Signal<H, Error>) -> Signal<(A, B, C, D, E, F, G, H), Error> {
-	return combineLatest(a, b, c, d, e, f, g)
```

Comment:
I started down that thinking.  

But this code is really the optimum sized memory and usage for the every situation.   

You could do everything with "max arrity" but then you use more memory than you need (for the typical 2,3,4 case). 

The reality is that while this may look like more "lines of code" it's the bare minimum number of actual operators needed to do any sized tuple up to max arrity.   

The current code  "looked smaller" but since each  version of the operation uses the version before, you are just satisfying some human need to see fewer lines of code.  Not thinking about the number of compiled instructions. 

I'm pretty sure this is the smallest amount of compiled code needed.   And will always use the smallest amount of memory.  

While maintaining support for both reference and value types, and maintaining type-correctness.  



Sent from my iPhone

> On Oct 30, 2015, at 4:34 PM, Neil Pankey <notifications@github.com> wrote:
> 
> In ReactiveCocoa/Swift/Signal.swift:
> 
> >  }
> >  
> >  /// Combines the values of all the given signals, in the manner described by
> >  /// `combineLatestWith`.
> >  @warn_unused_result(message="Did you forget to call `observe` on the signal?")
> >  public func combineLatest<A, B, C, D, E, F, G, H, Error>(a: Signal<A, Error>, _ b: Signal<B, Error>, _ c: Signal<C, Error>, _ d: Signal<D, Error>, _ e: Signal<E, Error>, _ f: Signal<F, Error>, _ g: Signal<G, Error>, _ h: Signal<H, Error>) -> Signal<(A, B, C, D, E, F, G, H), Error> {
> > -	return combineLatest(a, b, c, d, e, f, g)
> This is why I suggested combineLatestWith be implemented once at the max-arity. Then define the lower-arity overloads in terms of it. I've got it worked out in my head, but I haven't spent time actually trying to implement it. If I do this afternoon I'll open an alternative PR.
> 
> —
> Reply to this email directly or view it on GitHub.
>